### PR TITLE
[EH] Add new test setting for new Wasm EH

### PR DESCRIPTION
**Pull request by aheejin**
`with_all_eh_sjlj` and `with_all_sjlj` decorators (previously `with_both_eh_sjlj` and `with_both_sjlj` but recently got renamed) currently tests two modes: Emscripten EH and Wasm EH (pre-2023). We decided to switch to a new version of the Wasm EH proposal in Oct 2023, so this adds a new test setting for the new Wasm EH proposal in those decorators.

Currently we have two parameters:
- ``: Emscripten EH
- `wasm`: Wasm EH (pre-2023)

This changes it to these three parameters:
- `emscripten`: Emscripten EH
- `wasm`: Wasm EH (pre-2023)
- `wasm_exnref`: Wasm EH (New proposal adopted on Oct 2023)

To use the new mode in the command line, you use `-fwasm-exceptions` as in the old Wasm EH, but add `-sWASM_EXNREF` additionally.

This currently uses the Binaryen translator that translates old Wasm EH instruction to the new ones
(https://github.com/WebAssembly/binaryen/blob/main/src/passes/TranslateEH.cpp) at the end of the Binaryen pipeline to produce new binaries.

### CODE REVIEW

**Comment by aheejin** [comment]
The CI will pass once https://github.com/llvm/llvm-project/pull/91299 rolls into Chromium CI.

**Comment by dschuff** [comment]
Sorry to bikeshed the naming, but I think a name that describes the feature itself is better than just describing it in relation to the other feature (e.g. `wasm_exnref` instead of `new_wasm`)

**Comment by dschuff** [comment]
otherwise LGTM though.

File: test/common.py [hunk 1]
```diff
@@ -500,24 +510,29 @@ def metafunc(self, standalone):
 
 
 # Tests exception handling / setjmp/longjmp handling in Emscripten EH/SjLj mode
-# and if possible, new wasm EH/SjLj mode. This tests two combinations:
+# and new wasm EH/SjLj modes. This tests three combinations:
 # - Emscripten EH + Emscripten SjLj
-# - Wasm EH + Wasm SjLj
+# - Wasm EH + Wasm SjLj (Phase 3, to be deprecated)
+# - Wasm EH _ Wasm SjLj (New, experimental)
```
Comment (kripken) [inline test/common.py:516]: What does `_` mean here? (after "EH ")

**Comment by dschuff** [comment]
> > Sorry to bikeshed the naming, but I think a name that describes the feature itself is better than just describing it in relation to the other feature (e.g. `wasm_exnref` instead of `new_wasm`)
> 
> My thought was mostly that this new mode will be merged into `wasm` anyway so I didn't think too much about the name. But (whether they will be merged or not later) being specific in the name sounds good to me too. Do you think it is better to change the option name to `-sWASM_EXNREF` or something than the verbose `-sNEW_EXPERIMENAL_WASM_EXCEPTIONS`?

True; although the timeline for how long this option will have to exist is pretty unclear. Most users aren't going to want to use exnref until it has been supported in all browsers for quite a while, so even if exnref went to phase 4 tomorrow, I expect it to be at least a year until we would think about making exnref the default behavior for fwasm-exceptions. But in the meantime there will be 2 options that coexist. 
So... yeah, I think it probably does make sense to have a better name.

**Comment by aheejin** [comment]
Changed the option name to `-sWASM_EXNREF` and the test parameter postfix to `_wasm_exnref`.

**Review by kripken — APPROVED** [review]

**Review by dschuff — APPROVED** [review]

**Pull request merged**